File: Tests/swift-sdk-objc-tests/IterableInAppNotificationTests.m
//
//  IterableInAppNotificationTests.m
//  Iterable-iOS-SDK
//
//  Created by David Truong on 10/3/17.
//  Copyright © 2017 Iterable. All rights reserved.
//

#import <XCTest/XCTest.h>

@import IterableSDK;

@interface IterableInAppNotificationTests : XCTestCase

@end

@implementation IterableInAppNotificationTests

- (void)setUp {
    [super setUp];
    // Put setup code here. This method is called before the invocation of each test method in the class.
}

- (void)tearDown {
    // Put teardown code here. This method is called after the invocation of each test method in the class.
    [super tearDown];
}

- (void)testExample {
    // This is an example of a functional test case.
    // Use XCTAssert and related functions to verify your tests produce the correct results.
}

- (void)testPerformanceExample {
    // This is an example of a performance test case.
    [self measureBlock:^{
        // Put the code you want to measure the time of here.
    }];
}

- (void)testGetNextNotificationNil {
    NSDictionary *payload;
    NSDictionary *message = [IterableInAppManager getNextMessageFromPayload:payload];
    
    XCTAssertNil(message);
}

- (void)testGetNextNotificationEmpty {
    NSDictionary *payload = @{ @"inAppMessages" : @[] };
    NSDictionary *message = [IterableInAppManager getNextMessageFromPayload:payload];
    
    XCTAssertNil(message);
}

- (void)testNotificationCreation {
    //call showIterableNotificationHTML with fake data
    //Check the top level dialog
    
    NSString *htmlString = @"<a href=\"http://www.iterable.com\" target=\"http://www.iterable.com\">test</a>";
    
    IterableInAppHTMLViewController *baseNotification;
        baseNotification = [[IterableInAppHTMLViewController alloc] initWithData:htmlString];

    NSString *html = [baseNotification getHtml];
    XCTAssertEqual(html, htmlString);
}

- (void)testGetPaddingInvalid {
    NSDictionary *payload = @{};
    
    UIEdgeInsets insets = [IterableInAppManager getPaddingFromPayload:payload];
    
    XCTAssertTrue(UIEdgeInsetsEqualToEdgeInsets(insets, UIEdgeInsetsZero));
}

- (void)testGetPaddingFull {
    
    NSDictionary *payload = @{ @"top" : @{@"percentage" : @"0"}, @"left" : @{@"percentage" : @"0"}, @"bottom" : @{@"percentage" : @"0"}, @"right" : @{@"percentage" : @"0"}};
    
    UIEdgeInsets insets = [IterableInAppManager getPaddingFromPayload:payload];
    
    XCTAssertTrue(UIEdgeInsetsEqualToEdgeInsets(insets, UIEdgeInsetsZero));
    
    UIEdgeInsets padding = UIEdgeInsetsZero;
        padding.top = [IterableInAppManager decodePadding:[payload objectForKey:@"top"]];
        padding.left = [IterableInAppManager decodePadding:[payload objectForKey:@"left"]];
        padding.bottom = [IterableInAppManager decodePadding:[payload objectForKey:@"bottom"]];
        padding.right = [IterableInAppManager decodePadding:[payload objectForKey:@"right"]];

    XCTAssertTrue(UIEdgeInsetsEqualToEdgeInsets(insets, UIEdgeInsetsZero));
}

- (void)testGetPaddingCenter {
    
    NSDictionary *payload = @{ @"top" : @{@"displayOption" : @"AutoExpand"}, @"left" : @{@"percentage" : @"0"}, @"bottom" : @{@"displayOption" : @"AutoExpand"}, @"right" : @{@"percentage" : @"0"}};
    
    UIEdgeInsets insets = [IterableInAppManager getPaddingFromPayload:payload];
    
    XCTAssertTrue(UIEdgeInsetsEqualToEdgeInsets(insets, UIEdgeInsetsMake(-1, 0, -1, 0)));
    
    UIEdgeInsets padding = UIEdgeInsetsZero;
    padding.top = [IterableInAppManager decodePadding:[payload objectForKey:@"top"]];
    padding.left = [IterableInAppManager decodePadding:[payload objectForKey:@"left"]];
    padding.bottom = [IterableInAppManager decodePadding:[payload objectForKey:@"bottom"]];
    padding.right = [IterableInAppManager decodePadding:[payload objectForKey:@"right"]];
    
    XCTAssertTrue(UIEdgeInsetsEqualToEdgeInsets(insets, UIEdgeInsetsMake(-1, 0, -1, 0)));
}

- (void)testGetPaddingTop {
    
    NSDictionary *payload = @{ @"top" : @{@"percentage" : @"0"}, @"left" : @{@"percentage" : @"0"}, @"bottom" : @{@"displayOption" : @"AutoExpand"}, @"right" : @{@"percentage" : @"0"}};
    
    UIEdgeInsets insets = [IterableInAppManager getPaddingFromPayload:payload];
    
    XCTAssertTrue(UIEdgeInsetsEqualToEdgeInsets(insets, UIEdgeInsetsMake(0, 0, -1, 0)));
    
    UIEdgeInsets padding = UIEdgeInsetsZero;
    padding.top = [IterableInAppManager decodePadding:[payload objectForKey:@"top"]];
    padding.left = [IterableInAppManager decodePadding:[payload objectForKey:@"left"]];
    padding.bottom = [IterableInAppManager decodePadding:[payload objectForKey:@"bottom"]];
    padding.right = [IterableInAppManager decodePadding:[payload objectForKey:@"right"]];
    
    XCTAssertTrue(UIEdgeInsetsEqualToEdgeInsets(insets, UIEdgeInsetsMake(0, 0, -1, 0)));
}

- (void)testGetPaddingBottom {
    NSDictionary *payload = @{ @"top" : @{@"displayOption" : @"AutoExpand"}, @"left" : @{@"percentage" : @"0"}, @"bottom" : @{@"percentage" : @"0"}, @"right" : @{@"percentage" : @"0"}};
    
    UIEdgeInsets insets = [IterableInAppManager getPaddingFromPayload:payload];
    
    XCTAssertTrue(UIEdgeInsetsEqualToEdgeInsets(insets, UIEdgeInsetsMake(-1, 0, 0, 0)));
    
    UIEdgeInsets padding = UIEdgeInsetsZero;
    padding.top = [IterableInAppManager decodePadding:[payload objectForKey:@"top"]];
    padding.left = [IterableInAppManager decodePadding:[payload objectForKey:@"left"]];
    padding.bottom = [IterableInAppManager decodePadding:[payload objectForKey:@"bottom"]];
    padding.right = [IterableInAppManager decodePadding:[payload objectForKey:@"right"]];
    
    XCTAssertTrue(UIEdgeInsetsEqualToEdgeInsets(insets, UIEdgeInsetsMake(-1, 0, 0, 0)));
}

- (void)testNotificationPaddingFull {
    INAPP_NOTIFICATION_TYPE notificationType = [IterableInAppHTMLViewController setLocation:UIEdgeInsetsMake(0,0,0,0)];
    XCTAssertEqual(notificationType, INAPP_FULL);
}

- (void)testNotificationPaddingTop {
    INAPP_NOTIFICATION_TYPE notificationType = [IterableInAppHTMLViewController setLocation:UIEdgeInsetsMake(0,0,-1,0)];
    XCTAssertEqual(notificationType, INAPP_TOP);
}

- (void)testNotificationPaddingBottom {
    INAPP_NOTIFICATION_TYPE notificationType = [IterableInAppHTMLViewController setLocation:UIEdgeInsetsMake(-1,0,0,0)];
    XCTAssertEqual(notificationType, INAPP_BOTTOM);
}

- (void)testNotificationPaddingCenter {
    INAPP_NOTIFICATION_TYPE notificationType = [IterableInAppHTMLViewController setLocation:UIEdgeInsetsMake(-1,0,-1,0)];
    XCTAssertEqual(notificationType, INAPP_CENTER);
}

- (void)testNotificationPaddingDefault {
    INAPP_NOTIFICATION_TYPE notificationType = [IterableInAppHTMLViewController setLocation:UIEdgeInsetsMake(10,0,20,0)];
    XCTAssertEqual(notificationType, INAPP_CENTER);
}

@end

File: Tests/swift-sdk-swift-tests/IterableInAppNotificationTests.swift
//
//  IterableInAppNotificationTests.swift
//  swift-sdk-swift-tests

//  Created by David Truong on 10/3/17.
//  Migrated to Swift by Tapash Majumder on 7/10/18.
//  Copyright © 2018 Iterable. All rights reserved.
//

import XCTest

@testable import IterableSDK

class IterableInAppNotificationTests: XCTestCase {
    
    override func setUp() {
        super.setUp()
        // Put setup code here. This method is called before the invocation of each test method in the class.
    }
    
    override func tearDown() {
        // Put teardown code here. This method is called after the invocation of each test method in the class.
        super.tearDown()
    }

    func testGetNextNotificationNil() {
        let message = IterableInAppManager.getNextMessageFromPayload(nil)
        XCTAssertNil(message)
    }
    
    func testGetNextNotificationEmpty() {
        let message = IterableInAppManager.getNextMessageFromPayload([:])
        XCTAssertNil(message)
    }
    
    func testNotificationCreation() {
        //call showIterableNotificationHTML with fake data
        //Check the top level dialog
        
        let htmlString = "<a href=\"http://www.iterable.com\" target=\"http://www.iterable.com\">test</a>"
        let baseNotification = IterableInAppHTMLViewController(data: htmlString)
        let html = baseNotification.getHtml()
        XCTAssertEqual(html, htmlString)
    }
    
    func testGetPaddingInvalid() {
        let insets = IterableInAppManager.getPaddingFromPayload([:])
        XCTAssertTrue(UIEdgeInsetsEqualToEdgeInsets(insets, UIEdgeInsets.zero))
    }
    
    func testGetPaddingFull() {
        let payload: [AnyHashable : Any] = [
            "top" : ["percentage" : "0"],
            "left" : ["percentage" : "0"],
            "bottom" : ["percentage" : "0"],
            "right" : ["right" : "0"],
        ]
        
        let insets = IterableInAppManager.getPaddingFromPayload(payload)
        XCTAssertTrue(UIEdgeInsetsEqualToEdgeInsets(insets, UIEdgeInsets.zero))
        
        var padding = UIEdgeInsets.zero
        padding.top = CGFloat(IterableInAppManager.decodePadding(payload["top"]))
        padding.left = CGFloat(IterableInAppManager.decodePadding(payload["left"]))
        padding.bottom = CGFloat(IterableInAppManager.decodePadding(payload["bottom"]))
        padding.right = CGFloat(IterableInAppManager.decodePadding(payload["right"]))
        XCTAssertTrue(UIEdgeInsetsEqualToEdgeInsets(padding, UIEdgeInsets.zero))
    }
    
    func testGetPaddingCenter() {
        let payload: [AnyHashable : Any] = [
            "top" : ["displayOption" : "AutoExpand"],
            "left" : ["percentage" : "0"],
            "bottom" : ["displayOption" : "AutoExpand"],
            "right" : ["right" : "0"],
            ]
        
        let insets = IterableInAppManager.getPaddingFromPayload(payload)
        XCTAssertTrue(UIEdgeInsetsEqualToEdgeInsets(insets, UIEdgeInsets(top: -1, left: 0, bottom: -1, right: 0)))
        
        var padding = UIEdgeInsets.zero
        padding.top = CGFloat(IterableInAppManager.decodePadding(payload["top"]))
        padding.left = CGFloat(IterableInAppManager.decodePadding(payload["left"]))
        padding.bottom = CGFloat(IterableInAppManager.decodePadding(payload["bottom"]))
        padding.right = CGFloat(IterableInAppManager.decodePadding(payload["right"]))
        XCTAssertTrue(UIEdgeInsetsEqualToEdgeInsets(padding, UIEdgeInsets(top: -1, left: 0, bottom: -1, right: 0)))
    }
    
    func testGetPaddingTop() {
        let payload: [AnyHashable : Any] = [
            "top" : ["percentage" : "0"],
            "left" : ["percentage" : "0"],
            "bottom" : ["displayOption" : "AutoExpand"],
            "right" : ["right" : "0"],
            ]
        
        let insets = IterableInAppManager.getPaddingFromPayload(payload)
        XCTAssertTrue(UIEdgeInsetsEqualToEdgeInsets(insets, UIEdgeInsets(top: 0, left: 0, bottom: -1, right: 0)))
        
        var padding = UIEdgeInsets.zero
        padding.top = CGFloat(IterableInAppManager.decodePadding(payload["top"]))
        padding.left = CGFloat(IterableInAppManager.decodePadding(payload["left"]))
        padding.bottom = CGFloat(IterableInAppManager.decodePadding(payload["bottom"]))
        padding.right = CGFloat(IterableInAppManager.decodePadding(payload["right"]))
        XCTAssertTrue(UIEdgeInsetsEqualToEdgeInsets(padding, UIEdgeInsets(top: 0, left: 0, bottom: -1, right: 0)))
    }
    
    func testGetPaddingBottom() {
        let payload: [AnyHashable : Any] = [
            "top" : ["displayOption" : "AutoExpand"],
            "left" : ["percentage" : "0"],
            "bottom" : ["percentage" : "0"],
            "right" : ["right" : "0"],
            ]
        
        let insets = IterableInAppManager.getPaddingFromPayload(payload)
        XCTAssertTrue(UIEdgeInsetsEqualToEdgeInsets(insets, UIEdgeInsets(top: -1, left: 0, bottom: 0, right: 0)))
        
        var padding = UIEdgeInsets.zero
        padding.top = CGFloat(IterableInAppManager.decodePadding(payload["top"]))
        padding.left = CGFloat(IterableInAppManager.decodePadding(payload["left"]))
        padding.bottom = CGFloat(IterableInAppManager.decodePadding(payload["bottom"]))
        padding.right = CGFloat(IterableInAppManager.decodePadding(payload["right"]))
        XCTAssertTrue(UIEdgeInsetsEqualToEdgeInsets(padding, UIEdgeInsets(top: -1, left: 0, bottom: 0, right: 0)))
    }
    
    func testNotificationPaddingFull() {
        let notificationType = IterableInAppHTMLViewController.setLocation(UIEdgeInsets(top: 0, left: 0, bottom: 0, right: 0))
        XCTAssertEqual(notificationType, .full)
    }

    func testNotificationPaddingTop() {
        let notificationType = IterableInAppHTMLViewController.setLocation(UIEdgeInsets(top: 0, left: 0, bottom: -1, right: 0))
        XCTAssertEqual(notificationType, .top)
    }
    
    func testNotificationPaddingBottom() {
        let notificationType = IterableInAppHTMLViewController.setLocation(UIEdgeInsets(top: -1, left: 0, bottom: 0, right: 0))
        XCTAssertEqual(notificationType, .bottom)
    }

    func testNotificationPaddingCenter() {
        let notificationType = IterableInAppHTMLViewController.setLocation(UIEdgeInsets(top: -1, left: 0, bottom: -1, right: 0))
        XCTAssertEqual(notificationType, .center)
    }

    func testNotificationPaddingDefault() {
        let notificationType = IterableInAppHTMLViewController.setLocation(UIEdgeInsets(top: 10, left: 0, bottom: 20, right: 0))
        XCTAssertEqual(notificationType, .center)
    }
    
    func testDoNotShowMultipleTimes() {
        let shownFirstTime = IterableInAppManager.showIterableNotificationHTML("", callbackBlock: nil)
        let shownSecondTime = IterableInAppManager.showIterableNotificationHTML("", callbackBlock: nil)
        XCTAssertTrue(shownFirstTime)
        XCTAssertFalse(shownSecondTime)
    }
}
